File: app13_ruby_FastApi/backend/app/services/card_decode_error.rb
# frozen_string_literal: true

module CardDecodeError
  class Base < StandardError; end

  class UnrecognizedFields < Base
    def initialize(fields, context)
      super("unrecognized field(s) in #{context}: #{fields.join(', ')}")
    end
  end

  class MissingRequiredField < Base
    def initialize(field)
      super("missing required field: #{field}")
    end
  end

  class MissingCuratedSeverity < Base
    attr_reader :card_id

    def initialize(card_id)
      @card_id = card_id
      super("card '#{card_id}' is on a technical-threat deck but has no curated severity")
    end
  end

  class OrphanCurationEntry < Base
    attr_reader :card_id, :file

    def initialize(card_id, file)
      @card_id = card_id
      @file = file
      super("curation entry '#{card_id}' in #{file} has no matching card in its deck's raw YAML")
    end
  end

  class UnknownReference < Base
    attr_reader :value, :field, :card_id

    def initialize(value, field, card_id)
      @value = value
      @field = field
      @card_id = card_id
      super("card '#{card_id}' references unknown #{field} value '#{value}' (not in the bundled allowlist)")
    end
  end
end

File: app12_kotlin_android/data/src/main/kotlin/com/kotlinguard/data/cards/CardDecodeError.kt
package com.kotlinguard.data.cards

sealed class CardDecodeError(message: String) : Exception(message) {
    class MissingRequiredField(val field: String) : CardDecodeError("missing required field: $field")
    class OrphanCurationEntry(val cardId: String, val file: String) :
        CardDecodeError("curation entry '$cardId' in $file has no matching card in its deck's raw YAML")
    class MissingCuratedSeverity(val cardId: String) :
        CardDecodeError("card '$cardId' is on a technical-threat deck but has no curated severity")
    class UnknownReference(val value: String, val field: String, val cardId: String) :
        CardDecodeError("card '$cardId' references unknown $field value '$value' (not in the bundled allowlist)")
}
